Add contributing guidelines (#30)
EnergySystemsModellingLab/MUSE_2.0: CONTRIBUTING.md

that can be used to simulate change in an energy system over time. The goal of these guidelines is
to make the development of the project efficient and sustainable and to ensure that every commit
makes it better, more readable, more robust and better documented. Please, feel free suggest changes
and improvements.

(this guide is based on the [Atom editor guide](https://github.com/atom/atom/blob/master/CONTRIBUTING.md))

## Table Of Contents

[Code of Conduct](#code-of-conduct)

[How Can I Contribute?](#how-can-i-contribute)

- [Reporting Bugs](#reporting-bugs)
- [Suggesting Enhancements](#suggesting-enhancements)
- [Your First Code Contribution](#your-first-code-contribution)
- [Pull Requests](#pull-requests)

[Styleguides](#styleguides)

- [Git Commit Messages](#git-commit-messages)
- [Documentation Styleguide](#documentation-styleguide)

[Developer's setup](#developers-setup)

- [Continuous integration](#continuous-integration)

## Code of Conduct

This project and everyone participating in it is governed by the [MUSE Code of Conduct](CODE_OF_CONDUCT.md).
By participating, you are expected to uphold this code. Please report unacceptable behavior to

> (<ict-rse-team@imperial.ac.uk>.

Reviewer: Stray bracket:

```suggestion
<ict-rse-team@imperial.ac.uk>.
```

- **Describe the exact steps which reproduce the problem** in as many details as possible.
  For example, start by explaining how you installed MUSE and what you where trying to do.
- **Provide specific examples to demonstrate the steps**. Include links to files or GitHub projects,
  or copy/pasteable snippets, which you use in those examples. If you're providing snippets in the
  issue, use [Markdown code blocks](https://help.github.com/articles/markdown-basics/#multiple-lines).
- **Describe the behavior you observed after following the steps** and point out what exactly is the
  problem with that behavior.
- **Explain which behavior you expected to see instead and why.**
- **If there is any error output in the terminal, include that output with your report.**

Provide more context by answering these questions:

- **Did the problem start happening recently** (e.g. after updating to a new version of MUSE) or was
  this always a problem?
- If the problem started happening recently, **can you reproduce the problem in an older version of
  MUSE?** What's the most recent version in which the problem doesn't happen? You can download older

>   versions of MUSE from [the releases page](https://github.com/EnergySystemsModellingLab/MUSE_2.0/releases).

Reviewer: I guess it's fine to have this as a placeholder even though we don't have any releases yet...

- Engage the community in working toward the best possible MUSE
- Enable a sustainable system for MUSE's maintainers to review contributions

Please follow these steps to have your contribution considered by the maintainers:

1. **Describe clearly what is the purpose of the pull request**. Refer to the relevant issues on
   [Bugs](#reporting-bugs) or [Enhancements](#suggesting-enhancements). In general, an issue should
   always be open _prior_ to a pull request, to discuss its contents with a maintainer and make sure
   it makes sense for MUSE. If the pull request is a work in progress that will take some time to be
   ready but still you want to discuss it with the community, open a
   [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).
2. **Include relevant unit tests and integration tests, where needed**. MUSE's test suite is quite
   limited at the moment. We are working to improve this and tests as many features as possible, so
   any new addition to the code must come with its own set of tests to avoid going backwards in this
   matter.
3. **For new features and enhancements, include documentation and examples**. Both in the code, as

>    docstrings in classes, functions and modules, and as proper documentation describing how to use the

Reviewer: Let's rustify this:

```suggestion
   doc comments in structs, functions and modules, and as proper documentation describing how to use the
```

   new feature.
4. Follow the [styleguides](#styleguides)
5. After you submit your pull request, verify that all
   [status checks](https://help.github.com/articles/about-status-checks/) are passing

<details><summary>What if the status checks are failing?</summary>If a status check is failing, and
you believe that the failure is unrelated to your change, please leave a comment on the pull request
 explaining why you believe the failure is unrelated. A maintainer will re-run the status check for
 you. If we conclude that the failure was a false positive, then we will open an issue to track that
 problem with our status check suite.</details>

While the prerequisites above must be satisfied prior to having your pull request reviewed, the
reviewer(s) may ask you to complete additional design work, tests, or other changes before your pull
request can be ultimately accepted.

## Styleguides

### Git Commit Messages

- Use the present tense ("Add feature" not "Added feature")
- Use the imperative mood ("Move cursor to..." not "Moves cursor to...")
- Limit the first line to 72 characters or less
- Reference issues and pull requests liberally after the first line

> - When only changing documentation, include `[ci skip]` in the commit title

Reviewer: This doesn't apply:

```suggestion
```

- Consider starting the commit message with an applicable emoji:
  - :art: `:art:` when improving the format/structure of the code
  - :racehorse: `:racehorse:` when improving performance
  - :non-potable_water: `:non-potable_water:` when plugging memory leaks
  - :memo: `:memo:` when writing docs
  - :penguin: `:penguin:` when fixing something on Linux
  - :apple: `:apple:` when fixing something on macOS
  - :checkered_flag: `:checkered_flag:` when fixing something on Windows
  - :bug: `:bug:` when fixing a bug
  - :fire: `:fire:` when removing code or files
  - :green_heart: `:green_heart:` when fixing the CI build
  - :white_check_mark: `:white_check_mark:` when adding tests
  - :lock: `:lock:` when dealing with security
  - :arrow_up: `:arrow_up:` when upgrading dependencies
  - :arrow_down: `:arrow_down:` when downgrading dependencies

>   - :shirt: `:shirt:` when removing linter warnings

Reviewer: Do we care about the emoji thing? I don't mind either way.

(TIL there's an emoji for non-potable water :laughing:)

Author: Why not? I used to do this, but then I forgot, consumed by the dull task of committing stuff. The commit history in GitHub looks more colorful.

### Documentation Styleguide

- Use [Markdown](https://daringfireball.net/projects/markdown).
- Reference methods and classes in markdown with the custom `{}` notation:
  - Reference classes with `{ClassName}`
  - Reference instance methods with `{ClassName::methodName}`

>   - Reference class methods with `{ClassName.methodName}`

Reviewer: Cargo doc uses a different style, which we should probs link to: https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html

Might be worth signposting the rust documentation for this in general? https://doc.rust-lang.org/rustdoc/what-is-rustdoc.html

> ## Developer's setup
> 
> ### Continuous integration

Reviewer: I'm being pedantic here but... the other headings use title case:

```suggestion
## Developer Setup

### Continuous Integration
```

Pre-commit hooks are set up to run code quality checks (isort and black) before committing. To run
these locally, you will need to `pip install pre-commit` then `pre-commit install` - meaning you
need to have Python installed in your system and accessible. Now, quality assurance tools will be
run automatically with every commit.

Github workflows are set up to run the pre-commit actions and the tests automatically on every push

> action.

Reviewer: We also mention pre-commit in the readme, but I think it's good to have it here too.

The other thing we should probably mention is how to install the stable toolchain with `rustup`. It's also in the readme and you could just copy + paste it.

Author: I was going to, but then I thought it will be best to not duplicate information, otherwise, it is very easy to say contradictory things or to get out of sync. So, I would leave it in the README or moving it here, but not in both places. Maybe I can just add a link here to the README?

Reviewer: We could do. Alternatively we could just link from the readme to this section. Up to you which way round you'd rather do it!
